guidelines/auth.md

```diff
@@ -0,0 +1,126 @@
+# Authentication and Authorization Guideline
+
+This guideline documents the current best practices for authentication and
+authorization in Kubeflow.
+* Kubeflow Application Developers should use this guide to effectively integrate
+  their applications with Kubeflow.
+* Kubeflow Cluster Administrators should use this guide to understand Kubeflow's
+  auth architecture.
+
+
+## Introduction
+
+Kubeflow leverages Kubernetes to build its core APIs (CRDs), starting out with a
+few operators. The multi-tenancy story at this level was straightforward: direct
+interaction with the API via kubectl or other client, with the K8s API Server
+taking care of authentication and authorization.
+
+Soon after, Kubeflow introduced web apps to complement those core APIs and offer
+a better UX. This pushes the authentication and authorization problem out of the
+Kubernetes API Server and into the web apps.
+
+
+## Kubeflow Core (CRDs)
+Kubeflow started by introducing CRDs, providing an API on-top of Kubernetes.
+Interaction directly with the K8s API-Server via kubectl and
+authentication/authorization is a solved issue.
+
+
+### Authentication
+
+Whatever the [Kubernetes API-Server
+supports](https://kubernetes.io/docs/reference/access-authn-authz/authentication/):
+* ServiceAccount token
+* OIDC token
+* X509 Cert
+* Webhook
+* ...
+
+
+### Authorization
+
+K8s RBAC: Can `IDENTITY` perform `ACTION` on `RESOURCE` in `NAMESPACE`?
+
+
+## Kubeflow Web Apps
+
+Kubeflow wants to build on top of the functionality provided by its controllers.
+To provide a better UX, Kubeflow started building Web Apps to ease the use of
+its controllers. However, the story for authentication and authorization
+changes:
+* Authentication is performed by the Istio Gateway, not the K8s API Server.
+* Authorization is performed by the Web App Backend, not the K8s API Server.
+
+However, we still want to make sure that the user has the same identity and the
+same RBAC permissions are consulted.
+
+![kubeflow request flow](images/kubeflow_request_flow.svg)
+
+
+### Authentication
+
+Authentication is done at the Istio Gateway and the result is propagated to web
```

Comment:
Are we actually using the ISTIO gateway for authentication? At least in some cases I think Authentication (e.g. login) happens external to the gateway (e.g. IAP). In this case the gateway is just doing authorization; verifying user/system is authorized to access the service mesh.

Reply:
I see what you mean.
In both cases, you have a proxy of some kind that authenticates the user.
IAP is a Google IAM proxy that authenticates all requests with Google.
The AuthService turns the Gateway into an OIDC proxy/client, authenticating every request with an OIDC Provider.
The result is a JWT that the Istio Authentication Policy checks + some headers with certain values of the JWT.

Since, as you pointed out, this document is for application developers, I will trim the details.

```
How about, authentication (JWT vertification) is performed by Istio. The result is propagated to  web apps...
```

> In this case the gateway is just doing authorization

A small correction here, as authorization is always done server-side (at the workload, e.g., the Istio Sidecar of a Jupyter Notebook) and not client-side (at the gateway).

Reply:
I think we are doing Authorization in both places. At the gateway we provide a coarse grained policy (at least we do with GCP). So that policy basically says if the user isn't allowed to access any KF service then reject the request.  We then apply a finer grained policy in the istio side car.

Reply:
Got it! Addressing the original comment, I will amend the doc to say authentication is done at a proxy in front of the web-app. This is factually correct and generic enough.